File: bug42_accumulator_not_reset/buggy/main.py
import base64, struct

def run():
    groups = [[1,2,3],[4,5,6],[7,8,9],[10,11,12],[13,14,15,16,17,18,19,20]]
    group_hashes = []
    acc = 0
    for g in groups:
        for v in g:
            acc ^= v * 0x9E3779B97F4A7C15
            acc &= 0xFFFFFFFFFFFFFFFF
        group_hashes.append(acc)
    final = 0
    for i, h in enumerate(group_hashes):
        final ^= h * (i + 1)
        final = (final * 0xBF58476D1CE4E5B9) & 0xFFFFFFFFFFFFFFFF
    correct_hashes = []
    for g in groups:
        cacc = 0
        for v in g:
            cacc ^= v * 0x9E3779B97F4A7C15
            cacc &= 0xFFFFFFFFFFFFFFFF
        correct_hashes.append(cacc)
    correct_final = 0
    for i, h in enumerate(correct_hashes):
        correct_final ^= h * (i + 1)
        correct_final = (correct_final * 0xBF58476D1CE4E5B9) & 0xFFFFFFFFFFFFFFFF
    if final == correct_final:
        raise AssertionError("not resetting gives same final")
    raise ValueError(f"accumulator not reset between groups: got {final:#x} expected {correct_final:#x}")

run()

File: bug42_accumulator_not_reset/buggy/Main.java
import java.util.Base64;
import java.nio.ByteBuffer;

public class Main {
    static final long MAGIC1 = 0x9E3779B97F4A7C15L;
    static final long MAGIC2 = 0xBF58476D1CE4E5B9L;

    public static void main(String[] args) {
        int[][] groups = {
            {1,2,3}, {4,5,6}, {7,8,9}, {10,11,12}, {13,14,15,16,17,18,19,20}
        };

        /* BUG: acc is not reset between groups */
        long[] groupHashes = new long[groups.length];
        long acc = 0;
        for (int g = 0; g < groups.length; g++) {
            for (int v : groups[g]) {
                acc ^= (long)v * MAGIC1;
            }
            groupHashes[g] = acc;
        }

        long finalVal = 0;
        for (int i = 0; i < groupHashes.length; i++) {
            finalVal ^= groupHashes[i] * (long)(i + 1);
            finalVal = finalVal * MAGIC2;
        }

        long[] correctHashes = new long[groups.length];
        for (int g = 0; g < groups.length; g++) {
            long cacc = 0;
            for (int v : groups[g]) {
                cacc ^= (long)v * MAGIC1;
            }
            correctHashes[g] = cacc;
        }

        long correctFinal = 0;
        for (int i = 0; i < correctHashes.length; i++) {
            correctFinal ^= correctHashes[i] * (long)(i + 1);
            correctFinal = correctFinal * MAGIC2;
        }

        if (finalVal == correctFinal) {
            System.err.println("AssertionError: not resetting gives same final");
            System.exit(1);
        }
        System.err.printf("ValueError: accumulator not reset between groups: got 0x%x expected 0x%x%n",
                finalVal, correctFinal);
        System.exit(1);
    }
}
